### ISSUE TITLE
Nginx not updating upstream ip addresses when bosh changes them 

### ISSUE THREAD

**Issue by corlettb** [bosh-prometheus/prometheus-boshrelease#452]
What
----

In our deployment we often find that nginx needs to be restarted (or reloaded) to fix the upstream connections. This happens when bosh recreates nodes like prometheus or grafana and the upstream ip addresses change.

Apparently this is a known issue with nginx and is resolved [by the 'resolve' keyword](https://tenzer.dk/nginx-with-dynamic-upstreams/), only available in the paid for version. There is also a 'free' solution on the link.

Proposed Fixes
----

In order to fix this issue I have two solutions:

1. Write a service that would watch /etc/hosts and do a 'killall -s SIGHUP nginx' whenever it changed.

OR

2. Add bosh-dns to the node and make the following nginx config changes:
   - add resolver line to point to bosh-dns
   - change [grafana_location.conf](https://github.com/bosh-prometheus/prometheus-boshrelease/blob/master/jobs/nginx/templates/config/grafana_location.conf) to use the 'free' fix. The rendered config change would need to look something like:

```
proxy_pass            http://grafana/;
```

to

```
set $grafana_server http://7f61b995-070a-4e20-9143-d37b5c4a3a41.grafana.cf.prometheus.bosh:3000/;
proxy_pass            $grafana_server; 
```

Unfortunately there is no option to adjust grafana_location.conf in this way currently inside the template.

Questions
----

Are there any other alternative fixes that I'm missing? 

How would feel about a pull request to the grafana_location.conf template to allow the 'free' fix?

**Comment by benjaminguttmann-avtq** [comment]
Hi @corlettb , 
thanks for pointing this out.
I will try to reproduce the issue on our side to see if the mentioned fix that solve it, but in general we are open for Pull Requests to fix such an issue.

**Comment by benjaminguttmann-avtq** [comment]
Hi @corlettb I tried to reproduce the issue but was not yet able to do so.
What I tried to do was to adjust the static ip used in grafana to a new ip and re-deploy. According to BOSH the nginx VM has not been touched but when I try to access the Grafana Dashboard via Nginx it worked perfectly fine. Do you have any instruction on how to reproduce that issue.

**Comment by corlettb** [comment]
Hmmmm.

To reproduce the issue in our dev environment, I simulated ip changes by manually editing the /etc/hosts file on the nginx box. I then hit the grafana endpoint going via nginx and it still worked (expected a 502).

I executed a 'killall -s SIGHUP nginx' and nginx then started returning 502's.

Are you setting the resolver line in nginx.conf ?

**Comment by benjaminguttmann-avtq** [comment]
@corlettb we merged the PR of @dark5un which hopefully improves the situation for you as well. That PR is included in v28.0.0 which has been released yesterday. Feel free to re-open this ticket if your issue is not solved by this.
